sql/modules/Robot.sql

```diff
@@ -0,0 +1,126 @@
+-- Copyright (C) 2011 LedgerSMB Core Team.  Licensed under the GNU General
+-- Public License v 2 or at your option any later version.
+
+-- Docstrings already added to this file.
+
+BEGIN;
+
+CREATE OR REPLACE FUNCTION robot__get_my_entity_id() RETURNS INT AS
+$$
+        SELECT entity_id from users where username = SESSION_USER OR username = 'Migrator';
+$$ LANGUAGE SQL;
+
+COMMENT ON FUNCTION robot__get_my_entity_id() IS
+$$ Returns the entity_id of the current, logged in user.$$;
+
+DROP TYPE IF EXISTS robot_entity CASCADE;
+
+CREATE TYPE robot_entity AS (
+    entity_id int,
+    control_code text,
+    name text,
+    country_id int,
+    country_name text,
+    first_name text,
+    middle_name text,
+    last_name text,
+    entity_class int
+);
+
+CREATE FUNCTION robot__get(in_entity_id int)
+RETURNS robot_entity AS
+$$
+SELECT e.id, e.control_code, e.name, e.country_id, c.name,
+       p.first_name, p.middle_name, p.last_name, e.entity_class
+  FROM entity e
+  JOIN country c ON c.id = e.country_id
+  JOIN robot p ON p.entity_id = e.id
+ WHERE e.id = $1;
+$$ LANGUAGE SQL;
+
+CREATE FUNCTION robot__get_by_cc(in_control_code text)
+RETURNS robot_entity AS
+$$
+SELECT e.id, e.control_code, e.name, e.country_id, c.name,
+       p.first_name, p.middle_name, p.last_name, e.entity_class
+  FROM entity e
+  JOIN country c ON c.id = e.country_id
+  JOIN robot p ON p.entity_id = e.id
+ WHERE e.control_code = $1;
+$$ LANGUAGE SQL;
+
+DROP FUNCTION IF EXISTS robot__save (int, text, text, text, int);
+CREATE OR REPLACE FUNCTION robot__save
+(in_entity_id integer,
+in_first_name text, in_middle_name text, in_last_name text,
+in_country_id integer
+)
+RETURNS INT AS $$
```

Comment:
This is only the first draft for automation robots. It will have to evolve toward a more developed implementation with authorization, revoking, expiration, scope of action...